test/python/transpiler/test_unitary_synthesis_plugin.py

```diff
@@ -338,6 +339,100 @@ def test_target_overrides_basis_gates(self):
         self.assertEqual(call_kwargs["basis_gates"], set(target_basis_gates))
         self.assertNotEqual(call_kwargs["basis_gates"], set(standalone_basis_gates))
 
+    def test_fallback_on_default(self):
+        """Test that ``fallback_on_default`` works as expected."""
+        self.MOCK_PLUGINS["_controllable"].min_qubits = 0
+        self.MOCK_PLUGINS["_controllable"].max_qubits = np.inf
+        self.MOCK_PLUGINS["_controllable"].support([])
+
+        circ = QuantumCircuit(1)
+        circ.h(0)
+
+        qc = QuantumCircuit(1)
+        qc.unitary(Operator(circ).data, [0])
+
+        # The returned circuit consists of a single H-gate.
+        return_dag = circuit_to_dag(circ)
+
+        # Basis gates contain H and fallback is disabled:
+        # the default plugin should not be called.
+        with self.subTest("no fallback, synthesized circuit conforms"):
+            pm = PassManager(
+                [
+                    UnitarySynthesis(
+                        basis_gates=["t", "tdg", "h"],
+                        method="_controllable",
+                        fallback_on_default=False,
+                    )
+                ]
+            )
+            with unittest.mock.patch.object(
+                ControllableSynthesis, "run", return_value=return_dag
+            ) as plugin_mock:
+                with self.mock_default_run_method():
+                    pm.run(qc)
+                    plugin_mock.assert_called()
+                    self.DEFAULT_PLUGIN.run.assert_not_called()  # pylint: disable=no-member
+
+        # Basis gates do not contain H and fallback is disabled:
+        # the default plugin should not be called, even though the synthesized circuit
+        # does not conform to the target.
+        with self.subTest("no fallback, synthesized circuit does not conform"):
+            pm = PassManager(
+                [
+                    UnitarySynthesis(
+                        basis_gates=["t", "tdg"], method="_controllable", fallback_on_default=False
+                    )
+                ]
+            )
+            with unittest.mock.patch.object(
+                ControllableSynthesis, "run", return_value=return_dag
+            ) as plugin_mock:
+                with self.mock_default_run_method():
+                    pm.run(qc)
```

Comment:
What happens to the circuit in this case? Shouldn't transpilation fail and raise an error?

Reply:
That's also a good question. My view on this is that the default plugin is written by developers, and we assume that they know what they are doing (haha). They may cut corners like having the default synthesis plugin for transpilation into Clifford+T basis set call Solovay-Kitaev, which produces H, T, Tdg gates (and possibly not in the originally specified Clifford+T basis), but then they are responsible that the default pipeline calls `BasisTranslator` afterwards. What do you think?